dipy/reconst/shore.py

```diff
@@ -817,8 +817,8 @@ def shore_order(n, l, m):
         counter_i = 0
 
         counter = 0
-        for l_i in range(0, radial_order + 1, 2):
-            for n_i in range(l_i, int((radial_order + l_i) / 2) + 1):
+        for n_i in range(radial_order + 1):
+            for l_i in range(0, n + 1, 2):
                 for m_i in range(-l_i, l_i + 1):
                     if n == n_i and l == l_i and m == m_i:
                         counter_i = counter
```

Comment:
I know it's not your code, but I would ask the same question: can `counter_i` be found using simple math?

Reply:
Yes in both cases. This looping is mostly unnecessary and I've replaced it in a local commit that I haven't pushed yet. 

It turned out that in order to totally match BrainSuite I needed to add another spherical harmonics function.  Their definition of spherical harmonics (from here https://www.ncbi.nlm.nih.gov/pubmed/23353603) and their Merlet-compatible definition of the SHORE basis has some really nice properties when fitting real data. 

I haven't pushed these changes yet because their code is released under a GPLv2 license and that's not going to work with Dipy. Not sure what the best way forward is. The BrainSuite authors might agree to letting a python version of these two functions be released under a more permissive license. 

I also added the L1 optimization from the Merlet paper, which has performs very impressively.

Reply:
You have to register on their website and then you can download an archive of their matlab code. It's all GPLv2. @dvya is one of the authors

Reply:
Thanks! Unfortunately, you're right and the GPL is a non-starter for us :-( 

Unless what we have here is demonstrably wrong, I would recommend making this external to Dipy itself. You can grab our code and put it in a separate GPL-licensed repository. I'd be happy to help set that up.

One thing that I don't entirely understand: Is what we currently have flat-out wrong? Or is it really a difference of interpretation?

Reply:
It's not wrong, but also definitely not what's in the Merlet paper. At the minimum we might want change the docstrings so that they just reference the Zucchelli paper. I see differences in performance on synthetic datasets and the two methods have very different EAP response functions, with dipy's having considerably more artefact. 

If we get written permission from the authors to release a python implementation under a more permissive license, would that be ok?

Reply:
Hello - Sorry for chiming in so late (in between moves). So here is some technical background as to why we started this pull request in the first place.  All these basis functions are continuous functions so we have to truncate and approximate it while implementing. So what is currently in dipy is one approximation and it just leaves out more bases functions. And as Matt pointed out the reference is actually Zucchelli's paper. What BrainSuite has currently implemented is Merlet's l2 version which has more bases and we can show theoretically (was demonstrated at Brainhack 2018 - Varadarajan et al Neuroimage 2017) that it has better characteristics. (more accurate). 

So basically at Brainhack we discussed starting this pull request  and bringing this version to dipy.

Reply:
Regarding licenses - Can you please tell me exactly what information I need to get from BrainSuite to get this version of shore into dipy ?   If we cannot implement it for dipy then having a separate toolbox might not make sense because we have the implementation in BrainSuite - which is an open source toolbox. 

This is also important because I have a few of my methods implemented for dipy (testing ongoing), but these methods will be a part of next BrainSuite release.

Reply:
Hi all, and terribly sorry for being late to respond to this.  @mattcieslak and @dvya I do not think that we really need written permission as the same developer @dvya is providing the code to both software and the change is small. Also @dvya has already talked about this to the core devs of Brainsuite. If more algorithms are on the way then what Ariel suggests makes sense. 

Let me know if this is clear.

Reply:
Also @mattcieslak  definitely agree with @arokem with the class decision. Prefer to be one class if possible with two options new and old.